### ISSUE TITLE
Why the speed in minicaffe is much worse than caffe with same prototxt

### ISSUE THREAD

**Issue by yonghenglh6** [luoyetx/mini-caffe#48]
I compare the resnet from your run_test.cpp.  
But the performance is like below. The speed drop down than caffe.
Any ideas?
I'm useing the newest caffe with cudnn 5.1.5.

Thanks

![image](https://user-images.githubusercontent.com/6671814/30213047-ef6728f6-94d9-11e7-8581-307b3f3d5010.png)

**Comment by luoyetx** [comment]
would you mind to paste your test code?

**Comment by yonghenglh6** [comment]
In mini-caffe, I just add a "for(int i=0;i<100;i++)" before "test.Forward();" in run_net.cpp,  and get the time from your output.
In caffe, I use the caffe time, which is a mistake for memory use because it includes the backward process.

**Comment by yonghenglh6** [comment]
I get the memory use from "nvidia-smi" with watching in the flesh.

**Comment by yonghenglh6** [comment]
would u like to put up your performance with the resnet for a bug checking of mine?  Thank you.

**Comment by luoyetx** [comment]
I will test the network prototxt on 1070 later. With more details on mini-caffe and official caffe.

**Comment by yonghenglh6** [comment]
I use those code to test your every layer's time. But cannot find the reason.
![image](https://user-images.githubusercontent.com/6671814/30314182-863429e6-97d2-11e7-9f27-4e8480b1fb77.png)
And the result is:
![image](https://user-images.githubusercontent.com/6671814/30314198-949a3f52-97d2-11e7-9caa-9a4e4c989805.png)

**Comment by yonghenglh6** [comment]
Your net->Forward(2,3) give me a error.  So I can only use net->Forward(0,x) to get the time from the begin  and sub them.

**Comment by yonghenglh6** [comment]
As the net get longer, the performance diff between mini-caffe and caffe become larger, when I test them by adding layer step by step to construct a net.

**Comment by yonghenglh6** [comment]
Update the performance up.

**Comment by yonghenglh6** [comment]
I checked the cudnn and assured it ran well by adding some output info.

**Comment by luoyetx** [comment]
please refer to [profile.md](https://github.com/luoyetx/mini-caffe/blob/master/profile.md) to check the layer wise performance. I am writing tools to do the network benchmark.

**Comment by yonghenglh6** [comment]
I have tried Profile in the beginning, but the time shown in chrome is not consistent with my test result. However, it's most possible that I made the wrong usage way.

**Comment by luoyetx** [comment]
Pay attention to the Timer, it's not accuracy, use `Profiler::Now()` instead.

**Comment by luoyetx** [comment]
@yonghenglh6 You can try the `benchmark` branch, I modify the Profiler log, it now prints layer name not layer type. You can get same layer wise performance through `profile.json`

**Comment by luoyetx** [comment]
I find the performance is not stable under Windows platform, I will test on Linux later.

**Comment by yonghenglh6** [comment]
With your new benchmark tool, I found the bn layer is the main part that cause the difference, your bn is cost twice time than conv. So I changed to test it on google without bn layer and the result shows similar performance between caffe and mini-caffe.
Here is the details. By the way, my caffe use atlas lib, which I think it's not important.
![image](https://user-images.githubusercontent.com/6671814/30431119-2cdae960-9990-11e7-8bba-4c31977a8d35.png)

**Comment by luoyetx** [comment]
There is an optimization in this [commit](https://github.com/BVLC/caffe/commit/e93fcd267582888f960ca48d6e0c2e719d4ea09b) for BatchNorm Layer. I will update this from official caffe.

**Comment by yonghenglh6** [comment]
Everytime you request memory from pool, the blob will be in uninitial state, then it will call gpu_memset function in "to_gpu()", which will cost about 10% time more than original caffe whose blob will be keeped.

**Comment by luoyetx** [comment]
The default behave is the same in official Caffe [here](https://github.com/BVLC/caffe/blob/master/src/caffe/syncedmem.cpp#L70-L75).

**Comment by yonghenglh6** [comment]
Yes, but the official Caffe need not reallocate blob every forward and not call the function frequently. But your minicaffe keeping on setting the blob state to uninitial to reuse the memory and then rememset the memory when calling to_gpu(), which has caused a performance problem.
It can be optimized by designing the memset moment in layers carefully instead of doing it everytime "to_gpu()" called. 
In short words, this may be a problem of only minicaffe.

**Comment by luoyetx** [comment]
you mean [this function](https://github.com/luoyetx/mini-caffe/blob/master/src/syncedmem.cpp#L71)? It is a problem that this function called every time a new memory is requested. I think we can remove this function call, as the dirty data in the memory seems no problem for late use. What do you think?

**Comment by yonghenglh6** [comment]
Yes.
